### DISCUSSION TITLE
Listen on IPv6, also?

### DISCUSSION THREAD

**Discussion by troelsarvin** [tokio-rs/axum#834] [Q&A] [answered]
My server has both IPv4 and IPv6, and I want Axum to listen to both types of protocol on all interfaces.

The following results in an Axum which is available on port 3000 via IPv4 only. How can I make it available on IPv6, also?

```
    let addr = SocketAddr::from(([0, 0, 0, 0], 3000));
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
```

**Comment by SamuelBonilla** [comment, marked as answer]
Try with:

``` rust
let addr = ":::3000".parse().unwrap();
axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
```
 or 

``` rust
let addr = ":::3000".parse::<std::net::SocketAddr>().unwrap();
axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
```

**Reply by paolobarbolini** [reply]
I use `[::]:3000` and then if the application needs to access the client's address I try to run IPv6 addresses through [Ipv6Addr::method.to_ipv4_mapped](https://doc.rust-lang.org/nightly/std/net/struct.Ipv6Addr.html#method.to_ipv4_mapped) (since it's nightly only I copied the implementation into my code)

**Reply by troelsarvin** [reply]
Thanks, that works: But I think it's surprising that [::]:3000 implicitly also includes 0.0.0.0:3000. I wonder what I'd do, if I only wanted to bind to IPv6 :: and not involve IPv4 (although it's a rather hypothetical situation which is unlikely to be required in the real world).

**Reply by paolobarbolini** [reply]
> Thanks, that works: But I think it's surprising that [::]:3000 implicitly also includes 0.0.0.0:3000. I wonder what I'd do, if I only wanted to bind to IPv6 :: and not involve IPv4 (although it's a rather hypothetical situation which is unlikely to be required in the real world).

There's an IPv6 only flag for that. I'm not sure it can easily be set from Rust though

**Reply by jplatte** [reply]
I went down the rabbit hole, looks like it was briefly discussed (mostly bc. there's an inconsistency in defaults on Windows and Linux) but nothing happened in response 😥

See https://github.com/tokio-rs/tokio-core/issues/227 and the Rust PR linked from that.

The `socket2` crate (and `net2` before it) allows configuring this, although according to one comment in that tokio issue this wouldn't work on BSD which forces the compat stuff on. No idea about macOS. Plus hyper doesn't use that in `Server::bind` so it doesn't matter for this specific case anyways.

**Reply by troelsarvin** [reply]
@paolobarbolini, for the record: 
```
    let addr = ":::3000".parse::<std::net::SocketAddr>().unwrap();
```
does not work. It has to be the following way where the address part is encluded in square brackets:
```
    let addr = "[::]:3000".parse::<std::net::SocketAddr>().unwrap();
```

**Reply by borrow-checker** [reply]
Alternative if you like using constants not strings

`let addr = &SocketAddr::new(IpAddr::from(Ipv6Addr::UNSPECIFIED), 3000);`

Note that just like the string version `Ipv6Addr::UNSPECIFIED` refers to _both_ ipv4 and ipv6 in this context.

**Reply by llllvvuu** [reply]
Note that `::` / `Ipv6Addr::UNSPECIFIED` may or may not be ideal, since it's different from "both IPv4 and IPv6", it may also expose the service publicly: https://stackoverflow.com/a/72541698

**Reply by borrow-checker** [reply]
It's not different from both IPv4 and IPv6 as in Axum specifying [::] (which is equivalent to `Ipv6Addr::UNSPECIFIED`) then making a IpAddr with it assumes both v4 and v6 any.

**Reply by Havunen** [reply]
`let addr = "[::]:3000".parse::<std::net::SocketAddr>().unwrap();`   was a lifesaver... I was investigating random 2 second delays between azure front door and my rust axum backend after few days of debugging load balancers, tcp handshakes etc... I found out my program was not correctly listening IPv6 addresses which had been defined in DNS records with AAAA entry and Azure front door always silently retried in IPv4 format making all calls succeed but adding annoying random delays...

**Comment by jplatte** [comment]
I would imagine that this would work:

```rust
const PORT = 3000;

let svc = app.into_make_service();

let addr4 = SocketAddr::from((Ipv4Addr::UNSPECIFIED, PORT));
let addr6 = SocketAddr::from((Ipv6Addr::UNSPECIFIED, PORT));

let server4 = axum::Server::bind(&addr4).serve(svc.clone());
let server6 = axum::Server::bind(&addr6).serve(svc);

let (res4, res6) = futures_util::future::join(server4, server6).await;
res4.unwrap();
res6.unwrap();
```

**Reply by troelsarvin** [reply]
It doesn't work, which is odd:
```
    let app = Router::new().route("/", get(handler));

    const PORT: u16 = 3000;
    let addr_ipv4 = net::SocketAddr::from((net::Ipv4Addr::UNSPECIFIED, PORT));
    let addr_ipv6 = net::SocketAddr::from((net::Ipv6Addr::UNSPECIFIED, PORT));
    println!("addr_ipv4: {:?}", addr_ipv4);
    println!("addr_ipv6: {:?}", addr_ipv6);

    let svc = app.into_make_service();
    let server4 = axum::Server::bind(&addr_ipv4).serve(svc.clone());
    let server6 = axum::Server::bind(&addr_ipv6).serve(svc);
    let (res4, res6) = futures_util::future::join(server4, server6).await;
    res4.unwrap();
    res6.unwrap();
```
Results in:
```
addr_ipv4: 0.0.0.0:3000
addr_ipv6: [::]:3000
thread 'tokio-runtime-worker' panicked at 'error binding to [::]:3000: error creating server listener: Address already in use (os error 98)', /home/troels/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.14.17/src/server/server.rs:87:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I guess it's because binding to a :: IPv6 address has a strange side effect of also binding the 0.0.0.0 IPv4 interface. This is on Fedora Linux 35.

**Comment by troelsarvin** [comment]
By the way, in relation to this, https://serverfault.com/a/39561/15081 describes well how this is OS dependent, and on Linux behavior can be adjusted in /proc/sys/net/ipv6/bindv6only
Perhaps, if one wants exact control on Linux, one has to inspect /proc/sys/net/ipv6/bindv6only and act accordingly, but I think in practice almost all situations will benefit from just being able to bind "::" and get both IPv4 and IPv6.
